packages/datadog-instrumentations/src/cassandra-driver.js

```diff
@@ -0,0 +1,187 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const startCh = channel('apm:cassandra:query:start')
+const asyncEndCh = channel('apm:cassandra:query:async-end')
+const endCh = channel('apm:cassandra:query:end')
+const errorCh = channel('apm:cassandra:query:error')
+const connectionCh = channel(`apm:cassandra:general:connection`)
+
+addHook({ name: 'cassandra-driver', versions: ['>=3.0.0'] }, cassandra => {
+  shimmer.wrap(cassandra.Client.prototype, 'batch', batch => function (queries, options, callback) {
+    if (!startCh.hasSubscribers) {
+      return batch.apply(this, arguments)
+    }
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+    startCh.publish({ keyspace: this.keyspace, query: queries })
+
+    const lastIndex = arguments.length - 1
+    let cb = arguments[lastIndex]
+
+    if (typeof cb === 'function') {
+      cb = asyncResource.bind(cb)
+      arguments[lastIndex] = wrapCallback(asyncEndCh, errorCh, cb)
+    }
+
+    try {
+      const res = batch.apply(this, arguments)
+      if (typeof res === 'function' || !res) {
+        return wrapCallback(asyncEndCh, errorCh, res)
+      } else {
+        return res.then(
```

Comment:
The resolve and reject handlers should be bound explicitly, ideally to a single async resource instance.

Reply:
This also applies to other occurrences in this file.